lib/spaceship/client.rb

```diff
@@ -184,18 +193,43 @@ def login(user = nil, password = nil)
 
     def with_retry(tries = 5, &block)
       return block.call
-    rescue Faraday::Error::TimeoutError, AppleTimeoutError => ex # New Faraday version: Faraday::TimeoutError => ex
+    rescue Faraday::Error::ConnectionFailed, Faraday::Error::TimeoutError, AppleTimeoutError => ex # New Faraday version: Faraday::TimeoutError => ex
       unless (tries -= 1).zero?
         logger.warn("Timeout received: '#{ex.message}'.  Retrying after 3 seconds (remaining: #{tries})...")
-        sleep 3
+        sleep 3 unless defined?SpecHelper
+        retry
+      end
+      raise ex # re-raise the exception
+    rescue UnauthorizedAccessError => ex
+      if @loggedin
+        msg = "Auth error received: '#{ex.message}'. Login in again then retrying after 3 seconds (remaining: #{tries})..."
+        puts msg
+        logger.warn msg
+        do_login(self.user, @password)
+        sleep 3 unless defined?SpecHelper
+        debug_count_recovered
         retry
       end
-
       raise ex # re-raise the exception
     end
 
     private
 
+    def do_login(user, password)
+      @loggedin = false
+      ret = send_login_request(user, password) # different in subclasses
+      @loggedin = true
+      ret
+    end
+
+    # this mostly for troubleshooting while harderning the client
+    def debug_count_recovered
+      @recoveries_count = (@recoveries_count || 0) + 1
+      msg = "Succesfully recovered... (#{@recoveries_count} times!)"
+      puts msg
```

Comment:
Yes. I needed it to be able to monitor without digging into the fastlane log files. Will remove.